lib/interactivity-api.php

```diff
@@ -21,12 +21,22 @@ function gutenberg_reregister_interactivity_script_modules() {
 		$default_version
 	);
 
-	wp_register_script_module(
-		'@wordpress/interactivity-router',
-		gutenberg_url( '/build/interactivity/router.min.js' ),
-		array( '@wordpress/interactivity' ),
-		$default_version
-	);
+	if ( gutenberg_is_experiment_enabled( 'gutenberg-full-client-side-navigation' ) ) {
+		wp_register_script_module(
+			'@wordpress/interactivity-router',
+			gutenberg_url( '/build/interactivity/full-csn.min.js' ),
+			array( '@wordpress/interactivity' ),
+			$default_version
+		);
+		wp_enqueue_script_module( '@wordpress/interactivity-router' );
+	} else {
+		wp_register_script_module(
+			'@wordpress/interactivity-router',
```

Comment:
Calling a different script with the same name can cause misunderstandings if people use or extend the interactivity-router module. 
As is an experiment, we can keep going, but we will need to integrate csn into the library one received some feedback and testing.

Reply:
That makes sense 🙂 We should probably need to look for a better way once it is not an experiment.

Reply:
That's confusing to have two different files to play the same purpose. Would it be possible to have the functionality related to full client-side navigation in the existing router file but guarded with `if ( process.env.IS_GUTENBERG_PLUGIN ) {` check like in all other packages? That would simplify everything and better align with the rest of packages.

Reply:
> Would it be possible to have the functionality related to full client-side navigation in the existing router file but guarded with if ( process.env.IS_GUTENBERG_PLUGIN ) { check like in all other packages?

As this is an experiment, we wanted to avoid sending extra JS that is not needed in the current implementation.

Anyway, I'll think how this can be improved.

Reply:
Webpack, when it sees `process.env.IS_GUTENBERG_PLUGIN` set to `false` removes all the code behind this check. In effect, it won't impact the bundle size at all in WordPress core. In shouldn't matter that much in the Gutenberg plugin while running the experiment.

Reply:
Oh, I didn't know that 🙁 I can give that a try then. Although I am not sure yet how it should work because the conditional should not be when is Gutenberg but when the experiment is enabled.

Reply:
What is the difference in the size of the bundle?

Reply:
From a quick test, it seems the router would increase from 1,7 kb to 3.2 kb, more or less.

When you say "complicating the entire build system" do you mean the PHP conditional used here or creating a new entry in the interactivity webpack: [link](https://github.com/WordPress/gutenberg/pull/59707/files#diff-3d1911952fc568c4108afcd3bee2a8a5c5aea693d240df8a1a2812337256374dR22).

I believe we can look for alternatives to avoid both.

Reply:
It doesn't seem like a huge problem then given that the router is loaded asynchronously only when it's used. In effect, a user needs to hover the mouse over the link or click on it:

<img width="803" alt="Screenshot 2024-03-20 at 14 32 06" src="https://github.com/WordPress/gutenberg/assets/699132/399afd44-308b-4499-bfbd-e62605c4769d">


<img width="985" alt="Screenshot 2024-03-20 at 14 32 51" src="https://github.com/WordPress/gutenberg/assets/699132/44ffed1a-48e6-4f7c-a43b-995eb074b190">

Reply:
I've pushed another option I was considering:
* Keep the current scripts as they are: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/652a2a2de432dd132c364296565d99b5ca2df4a6).
* Register and load a new script for full client-side navigation when the experiment is enabled: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/442248c809b4eebffae92fbc1120a4a164f069ce).
* Load the original `@wordpress/interactivity-router` in the query loop only when the experiment is not enabled: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/0f3bc44fbe754f5cadd0957c93505078bfe7f22a).

@DAreRodz It'd be great to know your opinion about the best approach here.

Reply:
Huh, this one is tricky. Here's my two cents.

A pattern we thought about is that modules can define a light interface that is implemented by other async modules.

```jsx
import { store } from '@wordpress/interactivity';

store( 'somePlugin', {
	actions: {
		*doSomething() {
			// Let's assume `some-plugin/implementation` is a heavy module.
			yield import( 'some-plugin/implementation' );
			yield store( 'somePlugin/implementation' ).actions.doSomething();
		}
	}
} );
```

That approach doesn’t strictly apply here (we are already loading the `@wordpress/interactivity-router` module asynchronously), but my initial idea was to use that pattern to conditionally load the *region-based* or the *full-page* implementation of the router from `@wordpress/interactivity-router`, based on the `fullClientSideNavigation` config prop.

The immediate benefit is that blocks calling `store( 'core/router' ).actions.navigate` don’t have to be aware of which implementation should be used, or which module they have to import (see https://github.com/WordPress/gutenberg/pull/59707/commits/0f3bc44fbe754f5cadd0957c93505078bfe7f22a).

One side effect of this approach is that it would require defining two extra modules (two more entry points in Webpack and two more modules to register/enqueue in WP). I guess that’s *acceptable*?

When I discussed this with @SantosGuillamot a while ago, we saw that it was easier to just replace the module we send to the client when the `@wordpress/interactivity-router` is requested. That’s why we did so, although that’s probably not a good idea, as @gziolo mentioned. 😅

---

Anyways, as this is an experiment, I’m fine with any approach *for now*. We can iterate later on to find the best patterns for this use case.

Reply:
I don't mind implementing any approach, but it'd be great to make a decision so we can unblock this PR. We have to keep in mind that this is an experiment, so how the different modules will be handled can be explored deeper. The idea of having an experiment was precisely having it available without worrying too much about those aspects.

So it seems there are mainly three options at this point:

1. Include the conditional logic in the existing router

This increases the bundle size of the router even if the experiment is not enabled. The router is only loaded right now when the user hovers over a link in the query loop block.

On the other hand, it doesn't require changes in the webpack configuration.

2. Existing solution in this PR: Create two different routers, the existing one and a new one for full page client-side navigation

This requires creating a new module in WebPack, which is loaded when the experiment is enabled. When the experiment is enabled, it doesn't load the existing router for the query loop.

3. Explore the approach suggested by David [here](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1535479413)

I am not 100% sure how it differs from the solution implemented in this PR, but I could explore it more deeply.

---

@DAreRodz @gziolo Any strong opinions about the correct approach to unblock this experiment?

Reply:
The most important aspect, from my perspective, would be ensuring that the code doesn't diverge between two files. In particular, when a bug fix gets applied for the existing router, it's going to be very hard to ensure it gets mirrored in the other file. In effect, the simplest approach to unblock it here would be to move common logic to another file and import it from both entry points:

```
|-- interactivity-router
    |-- src
        |-- index.js // old entry point
        |-- experimental.js // experimental entry point
        |-- shared.js // shared logic, but can be also several files
```

Reply:
After taking a deeper look at it, I agree that most of the code of the router should be reused and having just one router checking the `navigationMode` seems the right approach to me. I've changed the PR to reflect that. Let me know what you think.

I've checked the bundle size and it increases from 1.7kB to 3.1kB. However, the code that handles the head tags is new and it will be needed for both modes in the future. Right now it is only used in this experiment but when it is tested and more mature, it should be used for region based client-side navigation as well.